File: C#/LinkedLists/LinkedList.cs
public class LinkedLists<T>
{
    Node<T>? head;
    Node<T>? tail;

    int len;

    public LinkedLists()
    {
        this.head = default(Node<T>);
        this.tail = default(Node<T>);
        this.len = 0;
    }

    //
    //
    // INSERT NODES
    //
    //
    public void InsertFirst(T data)
    {
        Node<T>? new_node;
        Node<T>? current_node;

        new_node = new Node<T>(data);
        if (this.head == null)
        {
            this.head = new_node;
            this.tail = new_node;
        }
        else
        {
            current_node = this.head;
            new_node.next = current_node;
            current_node.prev = new_node;
            this.head = new_node;
        }

        this.len++;
    }

    public void InsertLast(T data)
    {
        Node<T>? new_node;
        Node<T>? current_node;

        new_node = new Node<T>(data);
        if (this.head == null)
        {
            this.head = new_node;
            this.tail = new_node;
        }
        else
        {
            current_node = this.tail;
            new_node.prev = current_node;
            current_node.next = new_node;
            this.tail = new_node;
        }

        this.len++;
    }

    public void InsertMiddle(int index, T data)
    {
        Node<T>? new_node;
        Node<T>? prev_node;
        Node<T>? next_node;

        if (this.head == null && index > 0)
        {
            throw new NullReferenceException("List is empty, so index is out of bounds!");
        }
        if ((index > this.len && this.len > 1) || index < 0)
        {
            throw new IndexOutOfRangeException("Index is out of bounds!");
        }
        if (index == 0)
        {
            InsertFirst(data);
            return;
        }
        if (index == this.len)
        {
            InsertLast(data);
            return;
        }

        new_node = new Node<T>(data);

        int count = 0;

        prev_node = this.head;
        while (index - 1 > count)
        {
            prev_node = prev_node.next;
            count++;
        }

        next_node = prev_node.next;

        // Sets the next and previous for the new_node
        new_node.prev = prev_node;
        new_node.next = next_node;

        // Sets the new_node as next and prev for their respective next and prev nodes
        prev_node.next = new_node;
        next_node.prev = new_node;

        this.len++;
    }

    //
    //
    // REMOVE NODES
    //
    //
    public void RemoveFirst()
    {
        Node<T>? deleted_node;
        Node<T>? next_node;

        if (this.head == null)
        {
            throw new NullReferenceException("List is empty!");
        }

        if (this.len == 1) {
            this.head = null;
            this.tail = null;
        } else {
            deleted_node = this.head;
            next_node = deleted_node.next;
            deleted_node.next = null;
            next_node.prev = null;
            this.head = next_node;
        }

        this.len--;
    }

    public void RemoveLast()
    {
        Node<T>? deleted_node;
        Node<T>? prev_node;

        if (this.head == null)
        {
            throw new NullReferenceException("List is empty!");
        }

        if (this.len == 1)
        {
            this.head = null;
            this.tail = null;
        }
        else
        {
            deleted_node = this.tail;
            prev_node = deleted_node.prev;
            prev_node.next = null;
            deleted_node.prev = null;
            this.tail = prev_node;
        }

        this.len--;
    }

    public void RemoveFromIndex(int index)
    {
        Node<T>? deleted_node;
        Node<T>? prev_node;
        Node<T>? next_node;

        if (this.head == null)
        {
            throw new NullReferenceException("List is empty!");
        }
        if (index >= this.len || index < 0)
        {
            throw new IndexOutOfRangeException("Index is out of bounds!");
        }
        if (index == 0)
        {
            RemoveFirst();
            return;
        }
        if (index == this.len - 1)
        {
            RemoveLast();
            return;
        }

        prev_node = this.head;

        int count = 0;

        while (index - 1 > count)
        {
            prev_node = prev_node.next;
            count++;
        }

        deleted_node = prev_node.next;
        next_node = deleted_node.next;

        deleted_node.prev = null;
        deleted_node.next = null;

        next_node.prev = prev_node;
        prev_node.next = next_node;

        this.len--;
    }

    //
    //
    // GET NODES
    //
    //
    public T GetFirst()
    {
        if (this.head == null)
        {
            throw new NullReferenceException("List is empty!");
        }

        return this.head.value;
    }

    public T GetLast()
    {
        if (this.head == null)
        {
            throw new NullReferenceException("List is empty!");
        }

        return this.tail.value;
    }

    public T GetFromIndex(int index)
    {
        Node<T>? current_node;

        if (this.head == null)
        {
            throw new NullReferenceException("List is empty!");
        }
        if (index >= this.len || index < 0)
        {
            throw new IndexOutOfRangeException("Index is out of bounds!");
        }
        if (index == 0)
        {
            return GetFirst();
        }
        if (index == this.len - 1)
        {
            return GetLast();
        }

        current_node = this.head;

        int count = 0;

        while (index > count)
        {
            current_node = current_node.next;
            count++;
        }

        return current_node.value;
    }

    public int GetLength() 
    {
        return this.len;
    }

    // Returns list as string
    public string ReturnStringList()
    {
        Node<T>? current_node;

        if (this.head == null)
        {
            throw new NullReferenceException("List is empty!");
        }

        string? out_string = "";
        out_string += "[";

        current_node = this.head;
        out_string += current_node.value.ToString();
        while (current_node.next != null)
        {
            current_node = current_node.next;
            out_string += ", ";
            out_string += current_node.value.ToString();
        }

        out_string += "]";
        return out_string;
    }

    public List<object> ReturnActualList() 
    {
        Node<T>? current_node;
        List<object> out_list = new List<object>();

        if (this.head == null)
        {
            throw new NullReferenceException("List is empty!");
        }

        current_node = this.head;

        out_list.Add(current_node.value);

        while (current_node.next != null)
        {
            current_node = current_node.next;
            out_list.Add(current_node.value);
        }

        return out_list;
    }
}

File: Python/linked_list.py
from node import Node

class LinkedList:
    def __init__(self):
        """
        Initialize an empty LinkedList object with no set head or tail and a length of 0.
        """

        self.head = None
        self.tail = None
        self.length = 0

    # 
    # 
    # SETTERS
    # 
    # 
    def set_head(self, node: object) -> None:
        self.head = node

    def set_tail(self, node: object) -> None:
        self.tail = node

    # 
    # 
    # LENGTH MANIPULATION
    # 
    # 
    def increase_length(self) -> None:
        self.length += 1

    def decrease_length(self) -> None:
        self.length -= 1

    # 
    # 
    # INSERT NODES
    # 
    # 
    def insert_first(self, data: any) -> None:
        """
        Inserts Node at the front of the list.

        Args:
            data: any type of data you want in the node.
        """

        new_node = Node(data)

        if self.head is None:
            self.set_head(new_node)
            self.set_tail(new_node)
        else:
            new_node.set_next(self.head)
            self.head.set_prev(new_node)
            self.set_head(new_node)

        self.increase_length()

    def insert_last(self, data: any) -> None:
        """
        Inserts Node at the back of the list.

        Args:
            data: any type of data you want in the node.
        """

        new_node = Node(data)

        if self.head is None:
            self.set_head(new_node)
            self.set_tail(new_node)
        else:
            new_node.set_prev(self.tail)
            self.tail.set_next(new_node)
            self.set_tail(new_node)

        self.increase_length()

    def insert_middle(self, index: int, data: any) -> None:
        """
        Inserts Node at the stated index in the list.

        Args:
            index: the position of the new Node, 0 is the first position.
            data: any type of data you want in the node.
        Raises:
            IndexError: If the stated index is negative or bigger than the length of the list.
        """

        new_node = Node(data)

        if self.head is None and index > 0:
            raise IndexError("List is empty, index out of bounds")
        if (index > self.length and self.length > 0) or index < 0:
            raise IndexError("Index out of bounds")
        if index == 0:
            self.insert_first(data)
            return
        if index == self.length:
            self.insert_last(data)
            return

        count = 0
        prev_node = self.head

        while index-1 > count:
            prev_node = prev_node.next
            count += 1

        next_node = prev_node.next

        # Update next and prev
        new_node.set_next(next_node)
        new_node.set_prev(prev_node)

        prev_node.set_next(new_node)
        next_node.set_prev(new_node)

        self.increase_length()

    # 
    # 
    # REMOVE NODES
    # 
    # 
    def remove_first(self) -> None:
        """
        Removes the node at the front of the list.
        """

        if self.head is None:
            raise IndexError("List is empty")
        
        if self.length == 1:
            self.set_head(None)
            self.set_tail(None)
        else:
            deleted_node = self.head
            next_node = deleted_node.next

            deleted_node.set_next(None)
            next_node.set_prev(None)

            self.set_head(next_node)

        self.decrease_length()
        
    
    def remove_last(self) -> None:
        """
        Removes the node at the back of the list.
        """

        if self.head is None:
            raise IndexError("List is empty")
        
        if self.length == 1:
            self.set_head(None)
            self.set_tail(None)
        else:
            deleted_node = self.tail
            prev_node = deleted_node.prev

            deleted_node.set_prev(None)
            prev_node.set_next(None)

            self.set_tail(prev_node)

        self.decrease_length()
    
    def remove_from_index(self, index: int) -> None:
        """
        Removes the node at stated index of the list.

        Args:
            index: the index of the deleted Node, 0 is the first position.
        """

        if self.head is None and index > 0:
            raise IndexError("List is empty, index out of bounds")
        if (index > self.length and self.length > 0) or index < 0:
            raise IndexError("Index out of bounds")
        if index == 0:
            self.remove_first()
            return
        if index == self.length-1:
            self.remove_last()
            return

        prev_node = self.head

        count = 0

        while index-1 > count:
            prev_node = prev_node.next
            count += 1

        deleted_node = prev_node.next
        next_node = deleted_node.next

        # Update next and prev
        deleted_node.set_next(None)
        deleted_node.set_prev(None)

        prev_node.set_next(next_node)
        next_node.set_prev(prev_node)

        self.decrease_length()

    # 
    # 
    # GET NODE VALUES
    # 
    # 
    def get_first(self) -> any:
        """
        Returns the head value of the list.

        Return:
            Variable: value of the head Node.
        """

        if self.head is None:
            raise IndexError("List is empty")
        
        return self.head.value
    
    def get_last(self) -> any:
        """
        Returns the tail value of the list.

        Returns:
            Variable: value of the tail Node.
        """

        if self.head is None:
            raise IndexError("List is empty")
        
        return self.tail.value
    
    def get_value_from_index(self, index: int) -> any:
        """
        Returns the value of the Node from the stated index.

        Args:
            index: the index of the Node you want to get the value from, list starts at 0.

        Returns:
            Variable: value of the Node at the stated index.
        """

        if self.head is None and index > 0:
            raise IndexError("List is empty, index out of bounds")
        if (index > self.length and self.length > 0) or index < 0:
            raise IndexError("Index out of bounds")
        if index == 0:
            return self.get_first()
        if index == self.length-1:
            return self.get_last()
        
        current_node = self.head

        count = 0

        while index > count:
            current_node = current_node.next
            count += 1

        return current_node.value
    
    def __len__(self) -> int:
        """
        Returns the length of the list

        Returns:
            int: the length of the LinkedList
        """
        return self.length
    
    def __str__(self) -> str:
        """
        Returns an array of the values in the list

        Returns:
            str: a string looking like an array, not actual array.
        """

        if self.head == None:
            raise IndexError("List is empty")
        
        current_node = self.head

        out_string = "["
        out_string += str(current_node.value)

        while current_node.next is not None:
            current_node = current_node.next
            out_string += ", "
            out_string += str(current_node.value)

        out_string += "]"

        return out_string